tests/serialize_scene_tests.c

```diff
@@ -0,0 +1,191 @@
+    fputc(c, (FILE*)self_data);
+}
+void test_print_dbg(const char* c) {
+    printf("%s\n", c);
+}
+
+uint16_t test_file_read_char(void* self_data) {
+    return (uint16_t)fgetc((FILE*)self_data);
+}
+bool test_file_eof(void* self_data) {
+    return feof((FILE*)self_data) != 0;
+}
+
+typedef struct {
+    char* buffer;
+    unsigned int length;
+    unsigned int position;
+} stringsource;
+
+void test_string_write_buffer(void* self_data, uint8_t* buffer, uint16_t size) {
+    stringsource* ss = (stringsource*)self_data;
+    strncpy(ss->buffer + ss->position, (char*)buffer, size);
+    ss->position += size;
+    ss->length += size;
+}
+void test_string_write_char(void* self_data, uint8_t c) {
+    stringsource* ss = (stringsource*)self_data;
+    ss->buffer[ss->position] = c;
+    ss->position += 1;
+    ss->length += 1;
+}
+uint16_t test_string_read_char(void* self_data) {
+    stringsource* ss = (stringsource*)self_data;
+    if (ss->position < ss->length) {
+        char r = ss->buffer[ss->position];
+        ss->position += 1;
+        return r;
+    }
+    else { return -1; }
+}
+bool test_string_eof(void* self_data) {
+    stringsource* ss = (stringsource*)self_data;
+    return (ss->position >= ss->length);
+}
+
+tt_serializer_t test_file_writer, test_string_writer;
+tt_deserializer_t test_file_reader, test_string_reader;
+
+void init_serializers() {
+    test_file_writer.write_buffer = &test_file_write_buffer;
+    test_file_writer.write_char = &test_file_write_char;
+    test_file_writer.print_dbg = &test_print_dbg;
+
+    test_file_reader.read_char = &test_file_read_char;
+    test_file_reader.eof = &test_file_eof;
+    test_file_reader.print_dbg = &test_print_dbg;
+
+    test_string_writer.write_buffer = &test_string_write_buffer;
+    test_string_writer.write_char = &test_string_write_char;
+    test_string_writer.print_dbg = &test_print_dbg;
+
+    test_string_reader.read_char = &test_string_read_char;
+    test_string_reader.eof = &test_string_eof;
+    test_string_reader.print_dbg = &test_print_dbg;
+}
+
+void deserialize_fragment(char* fragment, scene_state_t* scene,
+                          char (*text)[SCENE_TEXT_LINES][SCENE_TEXT_CHARS]) {
+    stringsource ss;
+    ss.buffer = fragment;
+    ss.length = strlen(fragment);
+    ss.position = 0;
+
+    test_string_reader.data = (void*)&ss;
+
+    deserialize_scene(&test_string_reader, scene, text);
+}
+
+int compare_files(char* filename, FILE* a, FILE* b) {
+    fseek(a, 0, 0);
+    fseek(b, 0, 0);
+
+    int line = 1;
+
+    while (!feof(a) && !feof(b)) {
+        char ca = fgetc(a);
+        char cb = fgetc(b);
+        if (ca != cb) {
+            lprintf("At %s line %d, expected '%c', got '%c'", filename, line,
+                    ca, cb);
+            return -1;
+        }
+        if (ca == '\n') { line++; }
+    }
+
+    if (feof(a) != feof(b)) { return -1; }
+
+    return 0;
+}
+
+TEST test_round_trip_file(char* filename, char* tempfile) {
```

Comment:
Looking at the tests might be the fastest way to get an overview of what's going on here.